### PR TITLE
Multiprocessing logic

### PR DESCRIPTION
**Pull request by JoyboyBrian**

### CODE REVIEW

File: pyproject.toml [hunk 1]
```diff
@@ -38,6 +38,7 @@ classifiers = [
     "Programming Language :: Python :: 3.10",
     "Programming Language :: Python :: 3.11",
     "Programming Language :: Python :: 3.12",
+    "Programming Language :: Python :: 3.13",
```
Comment (zhiyuan8) [inline pyproject.toml]: please merge main

File: nexa/eval/evaluator.py [hunk 2]
```diff
@@ -159,117 +214,123 @@ def _adjust_config(task_dict):
...
         for instances in instances_by_doc_id.values():
             instances.sort(key=lambda x: x.idx)
+        if not task.instances:
+            continue  # Skip if no instances are present
+        if not task.instances[0].filtered_resps:
+            continue  # Skip if filtered_resps is empty
         for filter_key in task.instances[0].filtered_resps.keys():
             doc_iterator = task.doc_iterator(
-                rank=lm.rank, limit=task_limit, world_size=lm.world_size
+                rank=0, limit=task_limit, world_size=1
```
Comment (zhiyuan8) [inline nexa/eval/evaluator.py]: add comments and explain

File: nexa/eval/evaluator.py [hunk 3]
```diff
@@ -108,47 +116,94 @@ def _adjust_config(task_dict):
...
 
-        for x, req in zip(resps, cloned_reqs):
-            req.resps.append(x)
+        # Collect results and update progress bar
+        results_received = 0
+        total_results = len(requests)
+        while results_received < total_results:
+            try:
+                # Get result from result queue
+                idx, resp = result_queue.get(timeout=1)
```
Comment (zhiyuan8) [inline nexa/eval/evaluator.py:188]: what is queue.get() fails or timeout?

File: nexa/eval/evaluator.py [hunk 4]
```diff
@@ -108,47 +116,94 @@ def _adjust_config(task_dict):
...
     eval_tasks = get_task_list(task_dict)
 
     for task_output in eval_tasks:
         task: Task = task_output.task
         task_limit = get_sample_size(task, limit)
         task.build_all_requests(
             limit=task_limit,
-            rank=lm.rank,
-            world_size=lm.world_size,
+            rank=0,
```
Comment (zhiyuan8) [inline nexa/eval/evaluator.py]: add explanations here

File: nexa/eval/evaluator.py [hunk 5]
```diff
@@ -30,47 +32,55 @@
...
+            # Process the request
+            reqtype = req.request_type
+            resps = getattr(lm_worker, reqtype)([req])
+            # Put the response in the result queue
+            result_queue.put((idx, resps[0]))
+            task_queue.task_done()
+        except queue.Empty:
+            continue
+        except Exception as e:
+            # Print error to stderr (will be suppressed)
```
Comment (zhiyuan8) [inline nexa/eval/evaluator.py]: please improve this comment

**Review by zhiyuan8 — CHANGES_REQUESTED** [review]
Please ask David to test

**Review by zhiyuan8 — APPROVED** [review]

**Comment by JoyboyBrian** [comment]
Windows Issue 
```bash
Traceback (most recent call last):
  File "C:\nexa-sdk\nexa-sdk\nexa\gguf\nexa_inference_text.py", line 116, in _load_model
    self.model = Llama(
  File "C:\nexa-sdk\nexa-sdk\nexa\gguf\llama\llama.py", line 364, in __init__
    _LlamaModel(
  File "C:\nexa-sdk\nexa-sdk\nexa\gguf\llama\_internals_transformers.py", line 50, in __init__
    self.model = llama_cpp.llama_load_model_from_file(
OSError: exception: access violation reading 0x0000000000000000
2024-10-20:21:35:32,909 ERROR    [nexa_eval.py:145] An error occurred during evaluation: exception: access violation reading 0x0000000000000000
```

**Review by Davidqian123 — APPROVED** [review]

**Pull request merged**